src/compute/src/sink/continual_task.rs

```diff
@@ -0,0 +1,793 @@
+            return None;
+        };
+        // Reduce this down to one update per-time-per-worker before doing
+        // anything with it, so we don't waste work on unnecessarily high data
+        // volumes.
+        let ct_times = self
+            .ct_times
+            .iter()
+            .cloned()
+            .map(move |ct_times| times_reduce(name, ct_times));
+        let diff_times =
+            differential_dataflow::collection::concatenate(&mut first.scope(), ct_times);
+        Some(diff_times)
+    }
+}
+
+impl<G> SinkRender<G> for ContinualTaskConnection<CollectionMetadata>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    fn render_sink(
+        &self,
+        compute_state: &mut ComputeState,
+        _sink: &ComputeSinkDesc<CollectionMetadata>,
+        sink_id: GlobalId,
+        _as_of: Antichain<Timestamp>,
+        start_signal: StartSignal,
+        oks: Collection<G, Row, Diff>,
+        errs: Collection<G, DataflowError, Diff>,
+        append_times: Option<Collection<G, (), Diff>>,
+    ) -> Option<Rc<dyn Any>> {
+        let name = sink_id.to_string();
+
+        // SUBTLE: See the big module rustdoc for what's going on here.
+        let oks = step_forward(&name, oks);
+        // WIP we didn't step backward the errs so I don't think this is right.
+        let errs = step_forward(&name, errs);
+
+        let to_append = oks
+            .map(|x| SourceData(Ok(x)))
+            .concat(&errs.map(|x| SourceData(Err(x))));
+        let append_times = append_times.expect("should be provided by ContinualTaskCtx");
+
+        let write_handle = {
+            let clients = Arc::clone(&compute_state.persist_clients);
+            let metadata = self.storage_metadata.clone();
+            let handle_purpose = format!("ct_sink({})", name);
+            async move {
+                let client = clients
+                    .open(metadata.persist_location)
+                    .await
+                    .expect("valid location");
+                client
+                    .open_writer(
+                        metadata.data_shard,
+                        metadata.relation_desc.into(),
+                        UnitSchema.into(),
+                        Diagnostics {
+                            shard_name: sink_id.to_string(),
+                            handle_purpose,
+                        },
+                    )
+                    .await
+                    .expect("codecs should match")
+            }
+        };
+
+        let collection = compute_state.expect_collection_mut(sink_id);
+        let mut probe = ProbeHandle::default();
+        let to_append = to_append.probe_with(&mut probe);
+        collection.compute_probe = Some(probe);
+        let sink_write_frontier = Rc::new(RefCell::new(Antichain::from_elem(Timestamp::minimum())));
+        collection.sink_write_frontier = Some(Rc::clone(&sink_write_frontier));
+
+        // TODO(ct): Obey `compute_state.read_only_rx`
```

Comment:
> Though maybe when we split the persist_sink out, we just get this for "free"? It might just work out that this operator can send along the writes and they get buffered by the sink (dropping any that the shard passes) until it is allowed to write.

We're talking about the storage persist_sink, not the self-correcting one, right? Does that already have a way to observe the frontier of the output shard without doing appends itself?

Reply:
I haven't yet figured out if I think we can/should reuse the storage persist_sink out of the box, or take this as an opportunity to finally extract some common bits and make a `shard_sink` that lives in src/persist-client. But worst case, yeah, `WriteHandle::shared_upper` gets you the pubsub-updated latest upper of the shard (i.e. no crdb traffic) and `WriteHandle::wait_for_upper_past` can be used to find out when that upper changes (no crdb traffic in the common case of an upper advancing once per second)